middleware.ts

```diff
@@ -0,0 +1,23 @@
+import { NextResponse } from "next/server"
+import type { NextRequest } from "next/server"
+
+export function middleware(request: NextRequest) {
+  const path = request.nextUrl.pathname
+
+  const isPublicPath = path === "/login" || path === "/signup"
+
+  const token = request.cookies.get("token")?.value
+
+  if (isPublicPath && token) {
+    const response = NextResponse.redirect(new URL("/", request.nextUrl))
+    return response
+  }
+
+  if (path === "/" && !token) {
+    return NextResponse.redirect(new URL("/login", request.nextUrl))
+  }
+}
+
+export const config = {
+  matcher: ["/", "/login", "/signup"],
```

Comment:
this is pretty cool, i hadn't seen middleware implemented like this in `next.js` before